### DISCUSSION TITLE
[Question]: How to get several models in cvc5 base api?

### DISCUSSION THREAD

**Discussion by hagozaebii** [cvc5/cvc5#10166] [Q&A] [answered]
Hi, recently, I wanna get multiple models from one formula. For example, I can do what I wanna do in z3py as follows:
```
>>> from z3 import *
>>> solver = Solver()
>>> a = Int('a')
>>> b = Int('b')
>>> c = Int('c')
>>> p
a + a == b + c + a + b + c*c
>>> q = And(p, a != 0)
>>> q
And(a + a == b + c + a + b + c*c, a != 0)
>>> solver.add(q)
>>> solver.check()
sat
>>> solver.model()
[b = -4, c = 2, a = -2]
>>> solver.check()
sat
>>> solver.model()
[c = 1, a = -4, b = -3]
>>> solver.check()
sat
>>> solver.model()
[c = 3, b = -8, a = -4]
```
In contrast, cvc5 gives always the same model. If I wanna get new model different from the existing model that I got already, How can I get it in cvc5 base api?

Thanks for your time!

**Comment by Selebrator** [comment]
I too am interested in this. What I have done before is adding the negation of the previous models to the problem.

p = your problem
m = model of p
p' = p and not m
m' = model of p'

**Comment by mpreiner** [comment, marked as answer]
What you want is the block model functionality of cvc5: https://cvc5.github.io/docs/cvc5-1.0.8/api/cpp/solver.html#_CPPv4NK4cvc56Solver10blockModelEN5modes15BlockModelsModeE

You can either do this via the API or if the SMT2 front-end is used via the command `(block-model)`.

**Comment by hagozaebii** [comment]
@mpreiner To be honest, I wanna take more specific information as long as you are okay.

1. I am working in python base library. And `blockModel` method forces `mode` as an argument. I do not know what `mode` means exactly and why it is needed. if you are okay, could you let me know?
2. In my experience, I found no way to use several models simultaneously. I tried to deepcopy solver that has different model but it failed. Is there way to use several models simultaneously in cvc5 base lib in python3?

Thanks for your time!

**Reply by mpreiner** [reply]
1) You can find the documentation of the modes here: https://cvc5.github.io/docs-ci/docs-main/api/cpp/modes.html#_CPPv4N4cvc55modes15BlockModelsModeE

If you want to block the current model, i.e., the specific model values, then you have to use `BlockModelsMode::VALUES`.

2) If you query the values for all terms that you need via getValue(), and store them in a map, these values will not change when issuing the next checkSat() call.

**Reply by hagozaebii** [reply]
@mpreiner 

Thanks for your answer! If you are okay, I want to ask you something more. sorry for bothering you.

1. in python base API, what you said can be used in this way?
```
>>> from cvc5 import *
>>> solver = Solver()
>>> solver.setOption('produce-models', 'true')
>>> a = solver.mkConst(solver.getIntegerSort(), 'a')
>>> p = solver.mkTerm(Kind.DISTINCT, a, solver.mkInteger(1))
>>> solver.assertFormula(p)
>>> solver.checkSat()
sat
>>> solver.getValue(a)
0
>>> solver.blockModelValues(p)
>>> solver.checkSat()
sat
>>> solver.getValue(a)
(- 1)
```

2. What you said sounds like I need to save a model in Python data structure(e.g. dictionary) by myself. However, I think I cannot evaluate new formulas through `getValue` in Solver with my data structure. Is it possible to use `getValue` with my data structure(in other words, something that is not cvc5 object)? I want to evaluate new formulas not by substitution but a method `getValue` in cvc5 API.

Thanks for your time again.

**Reply by mpreiner** [reply]
1) What I meant was to use solver.blockModel() with the VALUES mode, then it will block the current model.

2) Yes, that's correct. In this case you can't use getValue(), but would have to evaluate it yourself.

**Comment by valeryjust** [comment]
Hi, @hagozaebii 
Sorry for bothering you.

I have tested z3 with cpp-api.


The models is same too, and are there more operation in z3?

**Reply by hagozaebii** [reply]
hi, i am sorry but i do not know as well.
in python api, i can take several models whenever I use `check` method.
